tech_reports/LLMs/llms.md

```diff
@@ -86,7 +87,231 @@ Authors:
+* Avoid pushing attention masks, rotation matrices and so on if they can be generated on-device or re-used between iterations
+
+In addition pay attention to where data is tilized and untilized. You don’t want to do this on the host! At the time of writing `to_torch` will by default do this on the host. You can untilize on-device like this:
+
+```python
+tt_out_tiled = tt_model(decode_input, current_pos, rot_mat=current_rot_mat)
+tt_out_row_major = ttnn.untilize(tt_out_tiled, use_multicore=True)
+tt_tok = ttnn.argmax(tt_out_row_major, dim=3, use_multicore=True)
+torch_tok = ttnn.to_torch(tt_tok)
+```
+
+Looking at host-device communications in a python profiler like `viztracer` is possible but be careful - when async-mode is on then any time spent in a communication call like `to_torch` can be comprised of up to three things:
+
+1. Time spent waiting for the device
+2. Time spent transferring data
+3. Time spent untilizing data
+
+If you want to measure the calls this way, turn async mode off. In this way the time your main python thread spends in `to_torch` will not include any of (1) and will be a closer approximation to what you probably wanted to measure.
+
+#### 4+5. Device dispatch and op performance
+
+This is the fun bit, but we need to do a little prep to get started. First, metal must be compiled with `-p` to enable device profiling:
+
+```bash
+./build_metal -p
+```
+
+Then we can record an op performance csv file with tracy. For the pytests, run it like this:
+
+```bash
+python -m tracy -r -p -v -m pytest path/to/test.py
+```
+
+This produces a file named something like `ops_perf_results_2024_11_01_15_33_18.csv` - that file is all we need from the profiler for now, but to learn more see the [Metal Profiler tech report](https://github.com/tenstorrent/tt-metal/blob/main/tech_reports/MetalProfiler/metal-profiler.md).
+
+> **Warning:** Only use one single trace execution step when profiling. At time of writing profiler support with tracing is still a work-in-progress and more iterations will result in a `AssertionError: Device data mismatch error`.
+
+> **Note:** If you see errors whilst running tracy, try this device-only profiling process instead: run with `TT_METAL_DEVICE_PROFILER=1 pytest path/to/test.py` and after the run completes run `tt_metal/tools/profiler/process_ops_logs.py --date` to generate the CSV file.
+
+This CSV file contains a wealth of information recorded from all the device during program execution. To summarize it into a more human-readable format we run the `perf_report.py` tool:
+
+```bash
+python models/perf/perf_report.py OPS_CSV_FILE
+```
+
+The [documentation for this tool](https://github.com/tenstorrent/tt-metal/tree/main/models/perf) describes how to use it to select specific ranges of operations. For device performance we strongly recommend looking at a single layer. You can do this by using `--id-range` or by changing your test to run only a single layer of the model (recommended).
```

Comment:
Where is the documentation for perf_report.py. The link only contains the source code for the scripts.